File: sdks/android/library/src/androidTest/java/org/xmtp/android/library/ServerConfigurationTest.kt
package org.xmtp.android.library

import androidx.test.ext.junit.runners.AndroidJUnit4
import kotlinx.coroutines.runBlocking
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test
import org.junit.runner.RunWith
import org.xmtp.android.library.messages.PrivateKeyBuilder

/**
 * Spec 006 CFG-106 for Android: one test reads every field of
 * [Client.serverConfiguration], one calls [Client.fetchServerConfiguration]
 * against the shared backend.
 *
 * The shared stack runs `dev/backend/local.toml`, so the identifier, the two
 * query limits, and the single anvil chain are asserted exactly. Every other
 * published value is a compiled default that an operator may retune, so those
 * are asserted as read and in range.
 */
@RunWith(AndroidJUnit4::class)
class ServerConfigurationTest : BaseInstrumentedTest() {
    /** CFG-080: the build snapshot, field by field. */
    @Test
    fun testServerConfigurationExposesEveryField() {
        val client = runBlocking { createClient(PrivateKeyBuilder()) }
        val configuration = client.serverConfiguration()

        assertEquals("org.xmtp.local", configuration.identifier)
        assertTrue(
            "server version should be published: ${configuration.serverVersion}",
            configuration.serverVersion.isNotBlank(),
        )
        // local.toml publishes no minimum, which leaves the deployment open to
        // every client version (CFG-005).
        assertEquals("", configuration.minLibxmtpVersion)

        val auth = configuration.auth
        assertFalse(auth.enabled)
        assertTrue(auth.keys.isEmpty())
        assertTrue(auth.audiences.isEmpty())
        assertTrue(auth.issuers.isEmpty())
        assertTrue(auth.requiredScopes.isEmpty())

        val retention = configuration.retention
        assertTrue(retention.groupMessageSeconds > 0L)
        assertTrue(retention.welcomeSeconds > 0L)
        assertTrue(retention.keyPackageSeconds > 0L)

        val limits = configuration.limits
        // The two values local.toml overrides.
        assertEquals(50L, limits.maxQueryLimit)
        assertEquals(50L, limits.defaultQueryLimit)
        assertTrue(limits.maxEnvelopeBytes > 0L)
        assertTrue(limits.maxRequestBytes > 0L)
        assertTrue(limits.maxResponseBytes > 0L)
        assertTrue(limits.maxPublishTopics > 0L)
        assertTrue(limits.maxQueryTopics > 0L)
        assertTrue(limits.maxNewestMetadataTopics > 0L)
        assertTrue(limits.maxNewestFullTopics > 0L)
        assertTrue(limits.maxUpdateAdds > 0L)
        assertTrue(limits.maxUpdateRemoves > 0L)
        assertTrue(limits.maxStreamTopics > 0L)
        assertTrue(limits.maxStaticTopics > 0L)
        assertTrue(limits.maxLookupIdentifiers > 0L)
        assertTrue(limits.maxScwSignatures > 0L)
        assertTrue(limits.maxIdentityEntries > 0L)
        assertTrue(limits.maxUpdateFramesPerSecond > 0)
        assertTrue(limits.maxUpdateBurst > 0)
        assertTrue(limits.maxPingFramesPerSecond > 0)
        assertTrue(limits.maxPingBurst > 0)

        val mls = configuration.mls
        assertTrue(mls.maxGroupMembers > 0L)
        assertTrue(mls.maxInstallationsPerInbox > 0L)
        // `[mls]` is absent from local.toml, so the flag reads as the published
        // default rather than as absent.
        assertEquals(true, mls.commitLogEnabled)

        // The one chain the local stack verifies, anvil.
        assertEquals(listOf("eip155:31337"), configuration.smartContractWalletChains)
    }

    /** CFG-081: the same values, with no database and no client. */
    @Test
    fun testFetchServerConfigurationWithoutAClient() {
        val fetched = runBlocking { Client.fetchServerConfiguration(localApi().backendUrl) }

        assertEquals("org.xmtp.local", fetched.identifier)
        assertTrue(fetched.serverVersion.isNotBlank())
        assertFalse(fetched.auth.enabled)
        assertEquals(50L, fetched.limits.maxQueryLimit)
        assertEquals(listOf("eip155:31337"), fetched.smartContractWalletChains)

        val client = runBlocking { createClient(PrivateKeyBuilder()) }
        assertEquals(client.serverConfiguration(), fetched)
    }

    /** CFG-082: a refresh returns the fetched value and leaves the snapshot alone. */
    @Test
    fun testRefreshServerConfigurationLeavesTheSnapshot() {
        val client = runBlocking { createClient(PrivateKeyBuilder()) }
        val snapshot = client.serverConfiguration()

        val refreshed = runBlocking { client.refreshServerConfiguration() }

        assertEquals(snapshot, refreshed)
        assertEquals(snapshot, client.serverConfiguration())
    }
}

File: sdks/browser/test/ServerConfiguration.test.ts
import { buildClient, createSigner } from "@test/helpers";
import type {
  AuthConfiguration,
  LimitsConfiguration,
  RetentionConfiguration,
} from "@xmtp/wasm-bindings";
import { describe, expect, it } from "vitest";

import { Client } from "@/Client";
import { fetchServerConfiguration } from "@/utils/serverConfiguration";

const backendUrl = import.meta.env.XMTP_BACKEND_URL as string;

const authFields: readonly (keyof AuthConfiguration)[] = [
  "enabled",
  "keys",
  "audiences",
  "issuers",
  "requiredScopes",
];

const retentionFields: readonly (keyof RetentionConfiguration)[] = [
  "groupMessageSeconds",
  "welcomeSeconds",
  "keyPackageSeconds",
];

// Spec 006 §5.2 publishes twenty limits. Every one is a JavaScript number.
const limitFields: readonly (keyof LimitsConfiguration)[] = [
  "maxEnvelopeBytes",
  "maxRequestBytes",
  "maxResponseBytes",
  "maxPublishTopics",
  "maxQueryTopics",
  "maxQueryLimit",
  "defaultQueryLimit",
  "maxNewestMetadataTopics",
  "maxNewestFullTopics",
  "maxUpdateAdds",
  "maxUpdateRemoves",
  "maxStreamTopics",
  "maxStaticTopics",
  "maxLookupIdentifiers",
  "maxScwSignatures",
  "maxIdentityEntries",
  "maxUpdateFramesPerSecond",
  "maxUpdateBurst",
  "maxPingFramesPerSecond",
  "maxPingBurst",
];

describe("ServerConfiguration", () => {
  it("should read every field of the server configuration snapshot", async () => {
    const { identifier } = createSigner();
    const client = await buildClient(identifier);
    const configuration = client.serverConfiguration();

    expect(Object.keys(configuration).sort()).toEqual(
      [
        "auth",
        "identifier",
        "limits",
        "minLibxmtpVersion",
        "mls",
        "retention",
        "serverVersion",
        "smartContractWalletChains",
      ].sort(),
    );

    expect(typeof configuration.identifier).toBe("string");
    expect(configuration.identifier).not.toBe("");
    expect(typeof configuration.serverVersion).toBe("string");
    expect(typeof configuration.minLibxmtpVersion).toBe("string");

    const { auth } = configuration;
    expect(Object.keys(auth).sort()).toEqual([...authFields].sort());
    expect(typeof auth.enabled).toBe("boolean");
    for (const key of auth.keys) {
      expect(Object.keys(key).sort()).toEqual(["alg", "kid"]);
      expect(typeof key.kid).toBe("string");
      expect(typeof key.alg).toBe("string");
    }
    for (const list of [auth.audiences, auth.issuers, auth.requiredScopes]) {
      expect(Array.isArray(list)).toBe(true);
      for (const entry of list) expect(typeof entry).toBe("string");
    }

    const { retention } = configuration;
    expect(Object.keys(retention).sort()).toEqual([...retentionFields].sort());
    for (const field of retentionFields) {
      // §7: uint64 reaches JavaScript as a number, never a bigint.
      expect(typeof retention[field]).toBe("number");
      expect(retention[field]).toBeGreaterThan(0);
    }

    const { limits } = configuration;
    expect(Object.keys(limits).sort()).toEqual([...limitFields].sort());
    for (const field of limitFields) {
      expect(typeof limits[field]).toBe("number");
      expect(limits[field]).toBeGreaterThan(0);
      expect(limits[field]).toBeLessThan(Number.MAX_SAFE_INTEGER);
    }

    const { mls } = configuration;
    expect(typeof mls.maxGroupMembers).toBe("number");
    expect(mls.maxGroupMembers).toBeGreaterThan(0);
    expect(typeof mls.maxInstallationsPerInbox).toBe("number");
    expect(mls.maxInstallationsPerInbox).toBeGreaterThan(0);
    // `commitLogEnabled` is optional: absent means the compiled default.
    expect(["boolean", "undefined"]).toContain(typeof mls.commitLogEnabled);

    const chains = configuration.smartContractWalletChains;
    expect(Array.isArray(chains)).toBe(true);
    for (const chain of chains) expect(typeof chain).toBe("string");
  });

  it("should fetch the server configuration without a client", async () => {
    const configuration = await fetchServerConfiguration(backendUrl);
    expect(configuration.identifier).not.toBe("");
    expect(typeof configuration.serverVersion).toBe("string");
    expect(typeof configuration.auth.enabled).toBe("boolean");
    expect(configuration.limits.maxEnvelopeBytes).toBeGreaterThan(0);

    const fromClient = await Client.fetchServerConfiguration({ backendUrl });
    expect(fromClient.identifier).toBe(configuration.identifier);
  });

  it("should refresh the server configuration without changing the snapshot", async () => {
    const { identifier } = createSigner();
    const client = await buildClient(identifier);
    const snapshot = client.serverConfiguration();

    const refreshed = await client.refreshServerConfiguration();
    expect(refreshed.identifier).toBe(snapshot.identifier);
    expect(client.serverConfiguration()).toEqual(snapshot);
  });
});
